### ISSUE TITLE
moving of uninitialized memory

### ISSUE THREAD

**Issue by oli-obk** [solson/miri#95]
right now miri can't do

```rust
pub struct Foo {
    inner: i32,
}

fn main() {
    unsafe {
        let foo = Foo {
            inner: std::mem::uninitialized(),
        };
        let bar = foo;
    }
}
```

but this is something the standard library does: https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/mutex.rs#L105

Not sure what the rules are behind that yet, but to the best of my knowledge UB is when you actually make a choice depending on undefined memory and not just copying undefined bytes around.

**Comment by solson** [comment]
This is actually a regression I caused with https://github.com/solson/miri/commit/244ae8eac7b4e0ea668f69feb25c525fb4b22281. :(

It only affects moves of uninitialized primitive types (for which `try_read_value` tries to read a value). I'm interested in pursuing a fix I discussed with @eddyb involving adding an `undefined` boolean flag to `PrimVal`, bringing it closer to being like a slice of an `Allocation`. Currently, `Allocation`s have bytes, relocations, and an undef mask, while `PrimVal` has only bytes and a relocation. I haven't investigated all the consequences of that change yet, though.
